crates/red_knot_test/src/matcher.rs

```diff
@@ -41,6 +41,23 @@ impl FailuresByLine {
     fn is_empty(&self) -> bool {
         self.lines.is_empty()
     }
+
+    pub(crate) fn offset_errors(self, offset: OneIndexed) -> Self {
+        Self {
+            failures: self.failures,
+            lines: self
+                .lines
+                .into_iter()
+                .map(|line_failures| LineFailures {
+                    line_number: line_failures
+                        .line_number
+                        .checked_add(offset)
+                        .expect("The number of lines in a test file should fit into a `usize`"),
+                    range: line_failures.range,
+                })
+                .collect(),
+        }
+    }
```

Comment:
I think we could avoid eagerly copying the entire `lines` vec by instead just adding a `starting_line_number` or `line_offset` field to `FailuresByLine`, and adding the offset to each line number in `FailuresByLine::iter` instead.